File: src/Other/HappyNumber.java
package Other;

import java.util.HashSet;
import java.util.Set;

// https://leetcode.com/problems/happy-number/
public class HappyNumber {
    public boolean isHappy(int n) {
        Set<Integer> inLoop = new HashSet<>();
        // it loops endlessly in a cycle so dup appears
        while (inLoop.add(n)) {
            int sum = getSquareSum(n);
            if (sum == 1) {
                return true;
            } else {
                n = sum;
            }
        }

        return false;
    }

    private int getSquareSum(int n) {
        int sum = 0, remain;
        while (n > 0) {
            remain = n % 10;
            sum += remain * remain;
            n /= 10;
        }
        return sum;
    }

    public static void main(String[] args) {
        HappyNumber h = new HappyNumber();
        int input = 19;
        System.out.println("is happy: " + h.isHappy(input));
    }
}

File: py/essential/HappyNumber.py
# https://leetcode.com/problems/happy-number/description/?envType=problem-list-v2&envId=rr2ss0g5
# Neet


class Solution:
    def isHappy(self, n: int) -> bool:
        def getSquareSum(input: int) -> int:
            squareSum = 0
            while input > 0:
                squareSum += (input % 10) ** 2
                input = input // 10
            return squareSum

        inLoop = set()
        inLoop.add(n)
        while True:
            squareSum = getSquareSum(n)
            if squareSum == 1:
                return True
            # it loops endlessly in a cycle so dup appears
            if squareSum in inLoop:
                return False
            inLoop.add(squareSum)
            n = squareSum


s = Solution()
print(s.isHappy(19))
print(s.isHappy(2))
print(s.isHappy(1))
